tower-retry/src/policies.rs

```diff
@@ -0,0 +1,61 @@
+use super::Policy;
+use futures_util::future;
+
+/// A very basic retry policy with a limited number of retry attempts.
+///
+/// FIXME: explain why not to use this.
+#[derive(Clone, Debug)]
+pub struct RetryLimit {
+    remaining_tries: usize,
+}
+
+impl RetryLimit {
+    /// Create a policy with the given number of retry attempts.
+    pub fn new(retry_attempts: usize) -> Self {
+        RetryLimit {
+            remaining_tries: retry_attempts,
+        }
+    }
+}
+
+impl<Req: Clone, Res, E> Policy<Req, Res, E> for RetryLimit {
+    type Future = future::Ready<Self>;
+    fn retry(&self, _: &Req, result: Result<&Res, &E>) -> Option<Self::Future> {
+        if result.is_err() {
```

Comment:
I wonder if there is a way to traitize this? For example, http retries you may want to retry 500s but not 400s. In this case, if I were to apply this to hyper I would only be able to retry system errors not http errors. What do you think about providing some method to allow users to specify what they want to retry?

Reply:
One possibility would be to construct the `RetryLimit` with a closure `F: FnMut(&E) -> bool`; then callers who wanted to retry all errors could construct it as, e.g.,
```
let policy = RetryLimit::new(2, |_| true);
```
and callers who want to filter retries can stick whatever logic they want in the closure.  (Does that seem like the right bound for the closure?)

Reply:
Though since this only returns an error that wouldn't support retrying on http::Responses?

I think we could use some trait _and_ a TraitFn adapter for a closure, what do you think?

Reply:
Hmm, I'm not sure what kind of trait you're thinking of. I think I'm missing something: if the user-supplied logic has access to the whole response and error, how would that be different from just implementing `Policy::retry` directly?

Reply:
This is closer to the related vector retrylogic thing I sent on discord, but the idea is that as a user I can just provide what i'd like to retry on instead of having to implement a retry policy myself.

Reply:
I wouldn't copy this 1-1 but https://github.com/timberio/vector/blob/master/src/sinks/util/retries.rs#L25

Reply:
> I wonder if there is a way to traitize this? For example, http retries you may want to retry 500s but not 400s.

For example, finagle uses a `ResponseClassifier` in a similar way.